pkg/console/auth-config/auth-config_test.go

```diff
@@ -0,0 +1,174 @@
+package auth_config
+
+import (
+	"crypto/x509"
+	"fmt"
+	"testing"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"k8s.io/client-go/util/cert"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/json"
+	framework "k8s.io/client-go/tools/cache/testing"
+)
+
+var _ = Describe("Auth config", func() {
+	var (
+		userHeaders        []string
+		groupHeaders       []string
+		extraPrefixHeaders []string
+
+		configMap *v1.ConfigMap
+
+		fakeSource *framework.FakeControllerSource
+		authReader Reader
+	)
+
+	BeforeEach(func() {
+		userHeaders = []string{"user-header-1", "user-header-2"}
+		groupHeaders = []string{"group-header-1", "group-header-2"}
+		extraPrefixHeaders = []string{"extra-prefix-header-1", "extra-prefix-header-2"}
+
+		fakeSource = framework.NewFakeControllerSource()
+
+		userHeadersJson, err := json.Marshal(userHeaders)
+		Expect(err).ToNot(HaveOccurred())
+
+		groupHeadersJson, err := json.Marshal(groupHeaders)
+		Expect(err).ToNot(HaveOccurred())
+
+		extraPrefixHeaders, err := json.Marshal(extraPrefixHeaders)
+		Expect(err).ToNot(HaveOccurred())
+
+		configMap = &v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      configMapName,
+				Namespace: configMapNamespace,
+			},
+			Data: map[string]string{
+				usernameHeadersKey:    string(userHeadersJson),
+				groupHeadersKey:       string(groupHeadersJson),
+				extraPrefixHeadersKey: string(extraPrefixHeaders),
+				clientCaKey:           testCa,
+			},
+		}
+
+		fakeSource.Add(configMap)
+
+		authReader, err = createReaderFromListerWatcher(fakeSource)
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	It("should read user headers from ConfigMap", func() {
+		userHeadersInConfig, err := authReader.GetUserHeaders()
+		Expect(err).ToNot(HaveOccurred())
+
+		Expect(userHeadersInConfig).To(Equal(userHeaders))
+	})
+
+	It("should fail if user headers are missing", func() {
+		delete(configMap.Data, usernameHeadersKey)
+		fakeSource.Modify(configMap)
+
+		Eventually(func() error {
+			_, err := authReader.GetUserHeaders()
+			return err
+		}, time.Second, 100*time.Millisecond).
+			Should(MatchError(ContainSubstring(usernameHeadersKey + " not found in configmap")))
+	})
+
+	It("should read group headers from ConfigMap", func() {
+		groupHeadersInConfig, err := authReader.GetGroupHeaders()
+		Expect(err).ToNot(HaveOccurred())
+
+		Expect(groupHeadersInConfig).To(Equal(groupHeaders))
+	})
+
+	It("should fail group headers are missing", func() {
+		delete(configMap.Data, groupHeadersKey)
+		fakeSource.Modify(configMap)
+
+		Eventually(func() error {
+			_, err := authReader.GetGroupHeaders()
+			return err
+		}, time.Second, 100*time.Millisecond).
+			Should(MatchError(ContainSubstring(groupHeadersKey + " not found in configmap")))
```

Comment:
Is this valid use of `MatchError` and `ContainSubstring`?

Reply:
Yes, it's valid. We use it in SSP for example:
https://github.com/kubevirt/ssp-operator/blob/3470569dbecbe6fee98b45b23a071d85c2cfae55/tests/webhook_test.go#L156

And here is the implementation of `MatchErrorMatcher.Match()`:
https://github.com/kubevirt/vm-console-proxy/blob/d139fecfbc715021b52ded2497f26c2993c539c5/vendor/github.com/onsi/gomega/matchers/match_error_matcher.go#L35-L42

It allows passing nested matchers.